### PR TITLE
Redial support

### PR DESCRIPTION
**Pull request by 20kdc**
Requires https://github.com/space-wizards/Robust.LoaderApi/pull/1 - Submodule update NOT here!

### CODE REVIEW

**Comment by 20kdc** [comment]
On the basis that SS14.Launcher has not had the submodule updated, and the PR is merged upstream, I'm doing the submodule update here.

File: SS14.Launcher/LauncherMessaging.cs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        catch (Exception)
+        {
+            // Ok, so we're server.
+            Console.WriteLine("We are primary launcher (or primary launcher is out for lunch)");
+        }
+        // Try to create server
+        try
+        {
+            _pipeServer = new NamedPipeServerStream(actualPipeName);
+            new Thread(() =>
```
Comment (PJB3005) [inline SS14.Launcher/LauncherMessaging.cs:68]: Using a separate thread for this is overkill. Just `Task.Run()` some async thing that uses async pipe server methods instead.

File: SS14.Launcher/LauncherMessaging.cs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        return false;
+    }
+
+    /// <summary>
+    /// Closes the pipe server remotely, which causes WaitForConnection to throw, which cleans up the pipe server thread.
+    /// This is important because otherwise the thread sticks around.
+    /// </summary>
+    public static void ShutdownPipeServer()
+    {
+        if (_pipeServer != null)
```
Comment (PJB3005) [inline SS14.Launcher/LauncherMessaging.cs:129]: Use cancellation tokens for this instead. Also maybe make sure the thread has exited cleanly?

File: SS14.Launcher/LauncherCommands.cs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+{
+    private static IDisposable? _timer;
+    private static string _reason = "";
+
+    /// <summary>
+    /// Starts the timer that handles commands and either defers them (connection commands when not ready) or handles them.
+    /// </summary>
+    public static void StartReceivingTimer(MainWindowViewModel windowVm, LoginManager loginMgr)
+    {
+        _timer = DispatcherTimer.Run(() =>
```
Comment (PJB3005) [inline SS14.Launcher/LauncherCommands.cs:25]: Don't use a timer for this. Post to the UI thread via `Dispatcher.UIThread` and either have the pipe thread do the command parsing itself or use a `Channel` of some kind for it (the former reduces complexity and is probably fine).

File: SS14.Launcher/LauncherCommands.cs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Avalonia.Threading;
+using Serilog;
+using SS14.Launcher.ViewModels;
+using SS14.Launcher.Models.Logins;
+
+namespace SS14.Launcher;
+
+public static class LauncherCommands
```
Comment (PJB3005) [inline SS14.Launcher/LauncherCommands.cs:15]: I don't like the fact that these are static.

File: SS14.Launcher/LauncherMessaging.cs [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipes;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace SS14.Launcher;
+
+public static class LauncherMessaging
```
Comment (PJB3005) [inline SS14.Launcher/LauncherMessaging.cs:13]: Ditto, no static pleas.e

File: SS14.Launcher/ViewModels/ConnectingViewModel.cs [hunk 6]
```diff
@@ -14,16 +14,19 @@ public class ConnectingViewModel : ViewModelBase
...
 
+    private string? _overrideReason;
+
     public bool IsErrored => _connector.Status == Connector.ConnectionStatus.ConnectionFailed ||
                              _connector.Status == Connector.ConnectionStatus.UpdateError ||
                              _connector.Status == Connector.ConnectionStatus.ClientExited &&
                              _connector.ClientExitedBadly;
 
-    public ConnectingViewModel(Connector connector, MainWindowViewModel windowVm)
+    public ConnectingViewModel(Connector connector, MainWindowViewModel windowVm, string? overrideReason)
```
Comment (PJB3005) [inline SS14.Launcher/ViewModels/ConnectingViewModel.cs:24]: I was thinking that instead of an override, this would be an *additional* line.

**Review by PJB3005 — CHANGES_REQUESTED** [review]

File: SS14.Launcher/LauncherMessaging.cs (hunk 1 above)
Comment (20kdc) [reply SS14.Launcher/LauncherMessaging.cs:68]: Does async work at pre-Avalonia-init time?

File: SS14.Launcher/LauncherCommands.cs (hunk 4 above)
Comment (20kdc) [reply SS14.Launcher/LauncherCommands.cs]: De-staticified

File: SS14.Launcher/LauncherMessaging.cs (hunk 5 above)
Comment (20kdc) [reply SS14.Launcher/LauncherMessaging.cs]: De-staticified also

File: SS14.Launcher/ViewModels/ConnectingViewModel.cs (hunk 6 above)
Comment (20kdc) [reply SS14.Launcher/ViewModels/ConnectingViewModel.cs]: Done.

File: SS14.Launcher/LauncherMessaging.cs (hunk 1 above)
Comment (20kdc) [reply SS14.Launcher/LauncherMessaging.cs]: Nevermind, found a solution to this, committed, pushed.

File: SS14.Launcher/LauncherMessaging.cs (hunk 2 above)
Comment (20kdc) [reply SS14.Launcher/LauncherMessaging.cs]: Done.

File: SS14.Launcher/LauncherCommands.cs (hunk 3 above)
Comment (20kdc) [reply SS14.Launcher/LauncherCommands.cs]: Not using a separate thread, so it's all fine now

**Pull request merged**